internal/k8shandler/reconciler.go

```diff
@@ -25,6 +29,41 @@ func (er *ElasticsearchRequest) L() logr.Logger {
 	return er.ll
 }
 
+func SecretReconcile(requestCluster *elasticsearchv1.Elasticsearch, requestClient client.Client) error {
+	var secretChanged bool
+
+	newSecretHash := getSecretDataHash(requestCluster.Name, requestCluster.Namespace, requestClient)
+	// compare the new secret with current one in the nodes
+	for _, node := range nodes[nodeMapKey(requestCluster.Name, requestCluster.Namespace)] {
+		if newSecretHash != node.getSecretHash() {
+			secretChanged = true
+
+			_, nodeStatus := getNodeStatus(node.name(), &requestCluster.Status)
+			nodeStatus.UpgradeStatus.ScheduledForCertRedeploy = corev1.ConditionTrue
+		}
+	}
+
+	if secretChanged {
+		// Cluster's secret has been updated, update the cluster status to be redeployed
+
+		nretries := -1
+		retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
```

Comment:
@ewolinetz , thanks for your suggestions. Regarding updating the cluster status, shall I do a deepcopy before sending the update API request, something like 

```go
clusterStatus := requestCluster.Status.DeepCopy()

// update the status in the copied version
_, nodeStatus := getNodeStatus(node.name(), clusterStatus)
nodeStatus.UpgradeStatus.ScheduledForCertRedeploy = corev1.ConditionTrue

// Get the latest cluster status
if err := er.client.Get(context.TODO(), types.NamespacedName{Name: cluster.Name, Namespace: cluster.Namespace}, requestCluster); err != nil {
    return err
}
requestCluster.Status = clusterStatus
requestClient.Status().Update(context.TODO(), requestCluster)
```

Please advise.

Reply:
i don't think its necessary to do a deep copy... however within the retry loop we should be doing the following operations (and I think the deepcopy made this easier for us in the past... we would make a DeepCopy and then make the desired changes to it and check it against what currently exists out there):

1) get the current value of status
2) check if we need to change what is currently out there vs what we want to set (e.g. corev1.ConditionTrue for ScheduledForCertRedeploy)
3) update those values if necessary
4) perform the Status().Update() with this updated status

Reply:
You are right. According to the [doc of retry](https://github.com/kubernetes/client-go/blob/b72204b2445de5ac815ae2bb993f6182d271fdb4/util/retry/util.go#L70-L71), the callback fn should get the current status first.
I moved 1) the get current status, 2) check if status needs to be changed, and 3) update status into the retry loop. Please let me know if that is correct.